TFT/src/User/Menu/FeatureSettings.c

```diff
@@ -383,7 +383,7 @@ void menuFeatureSettings(void)
         listViewRefreshItem(index);
       }
 
-    loopProcess();
+    loopProcessWithPopup();
```

Comment:
It is better to keep the old name `loopProcess()` for when it is expected to run normally.

Reply:
workin' on it already

Reply:
...and done

Reply:
I sugested to keep the old name.

Reply:
the `_MenuLoop()` is not required. it just makes the name longer and confusing.

Reply:
> It is better to keep the old name `loopProcess()` for when it is expected to run normally.

When is it expected to run normally? Define "normally" please.


> the `_MenuLoop()` is not required. it just makes the name longer and confusing.

`loopProcess_MenuLoop()` is not suitable for any other place than the menu's main while loop (at least for this moment), as its name suggests it.
`loopProcess()` is a generic subroutine whilst `loopProcess_MenuLoop()` is one with a specific usage that uses `loopProcess()` as other function uses that too.
I know you never agreed with me ever, so if you insist, names can be changed to whatever you wish. There are many ways to skin a cat.

Reply:
@kisslorand I agree with Gurmeet for different reasons:
1) maintain loopProcess where it is currently used.
  1) In case the menu using loopProcess is stateless, there is no problem on reloading the menu
  2) In case the menu using loopProcess is statefull (e.g. Z Offset, Mesh Edit etc...) there is no problem on reloading the menu
2) your proposed loopProcess_MenuLoop on the menu will not redraw the menu in case a popup is displayed and handled by loopProcess_MenuLoop
3) the use of macro in Extrude and LoadUnload menu is absolutely not need and vanish the goal to handle the popup without the need to reload the Menu

Point 2) and 3) can be easily satisfied by a function with the same logic of loopProcessToContiotion with the exception of loop on the condition:

**bool loopProcessWithPopup(bool reloadMenuOnPopup);**

RETURN:
  * true: if a popup was handled
  * false if no popup was handled

INPUT:
  * reloadMenuOnPopup: taken into account in case a popup was handled:
    * false: to let the caller handle the menu relaod. Usable in case the caller is a menu so it can decide to simply redraw the menu instead of reloading the entire menu
    * true: used in case the caller is not a menu so it is not able to provide any kind of menu relaod (e.g. mustStoreCmd).  The previous menu will be forced to be reloaded

That function can be used on both a menu (e.g. LoadUnload) and a not menu function (such as mustStoreCmd).
If used on menu (e.g. LoadUnload), the caller must check the return value and in case "true" was returned it can simply readraw the GUI. E.g. on LoadUnload menu:

if (loopProcessToCondition(&isNotEmptyCmdQueue, false) == true)  // wait for the communication to be clean
  menuDrawPage(&loadUnloadItems);

if (loopProcessWithPopup(false) == true)
  menuDrawPage(&loadUnloadItems);

If used on a not menu the caller does not need to check the return value. E.g. on mustStoreCmd:

while (infoCmd.count >= CMD_MAX_LIST)
{
  loopProcessWithPopup(true);  // wait for a free slot in the queue in case the queue is currently full
}

Reply:
Point 2:
You might missed the changes in Main.c. 
`loopProcess_MenuLoop()` triggers the menu exit and the redraw of it in one single step (as opposed to two steps as before).

Point 3:
That macro is there to let the user know the TFT is busy with something. Without it if a popup appears and user selects an action (OK, Cancel) than the popup will hang there until the loop is finished. 
I know it is not needed from a functionality point of view. It's there foe smoother user experience (as so many other things in this FW).

Reply:
> The best suggestion I can give is simply to let me provide the ehnamcements on loppProcessToCondition and loopProcessWithPopup, and then you can use them in your new or existing menus according to your needs

The solution in this PR is universal, no need to make new functions if new loops appear with new conditionals. It covers now all situations.

Want a loop without popup handling? You have it, it's the old loopProcess().
Want a loop that handles popups? You have it, it's the loopProcess_PopupHandle().
Want a clean exit from a menu and redraw the menu in one single shot in case there was a popup? You have it, it's the  loopProcess_MenuLoop().
Want a flag that there has been a popup? You have it, loopProcess_PopupHandle() will set it by the value of "lastMenu". If its value is "NULL" there has been a  handled popup.
Want a flag that there has been a popup and a refresh is needed? You have it, loopProcess_MenuLoop() will set the value of  "infoMenu.menu[infoMenu.cur]" to "NULL".

Reply:
> > It is better to keep the old name `loopProcess()` for when it is expected to run normally.
> 
> When is it expected to run normally? Define "normally" please.

Ok, when its function does not change after merging the changes (which is: it still performs the old routine), then it is better to leave the name as is it. 

> `loopProcess_MenuLoop()` is not suitable for any other place than the menu's main while loop (at least for this moment), as its name suggests it. 
> `loopProcess()` is a generic subroutine whilst `loopProcess_MenuLoop()` is one with a specific usage that uses `loopProcess()` as other function uses that too.

There are already functions For these purposes (`loopBackEnd()` for handling background tasks, `loopFrontEnd()` for handling UI related tasks) but instead of improving on them, more nested functions are being created for the same purpose which will make it hard to track bugs consume more CPU cycles. 

> I know you never agreed with me ever, so if you insist, names can be changed to whatever you wish. There are many ways to skin a cat.

I do not insist. I gave a suggestion. You take it as you like. I assume this is a discussion. It is not a discussion if everyone is having the same opinion. Members provide their views on my PRs also and many of them do not agree with many things. This does not mean I should start complaining about it. Stop acting like a wife. Why do you always become offended as soon as I comment on anything?
why would you skin a cat? I do not know who created this line but that is so cruel!

Reply:
> Ok, when its function does not change after merging the changes (which is: it still performs the old routine), then it is better to leave the name as is it.

The `loopProcess_MenuLoop()` differs in its function from `loopProcess()`. Their purpose is the same but not their function.


> There are already functions For these purposes (`loopBackEnd()` for handling background tasks, `loopFrontEnd()` for handling UI related tasks) but instead of improving on them, more nested functions are being created for the same purpose which will make it hard to track bugs consume more CPU cycles.

Initially I made it with macros to avoid nesting but I switched to functions because I had a hunch you would disagree (with the macros). I used macro only where I really do not had any other choice.

 
>Stop acting like a wife. Why do you always become offended as soon as I comment on anything?

It's OK, let it out. I guess, after all, it is  a friendly discussion, right?
As far as I remember I took advices from you without "become offended" and most probably will do it  again.

> why would you skin a cat? I do not know who created this line but that is so cruel!

Some say it was Seba Smith in 1854  in the work ’Way down East or, Portraitures of Yankee Life".
You might like the lighter version: "there's more than one way to peel an orange"